### PR TITLE
CP-12640: Generate alert when log partition is full

### PR DESCRIPTION
**Pull request by sharady**
Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

### CODE REVIEW

**Comment by johnelse** [comment]
This implementation is very closely tied to the output format of `df` - why not use [os.statvfs](https://docs.python.org/2/library/os.html#os.statvfs) instead?

I believe you can use `f_bsize * f_bfree` to calculate the free space.

**Comment by thomassa** [comment]
@johnelse makes a valid point, though we were doing this already in the existing function `ccent_fs_usage` which calls out to `df /etc/passwd` to lean the free space on the main filesystem.

In the new function `get_percent_log_fs_usage` why does it use `df /dev/sda5` rather than `df /var/log`? Choosing a specific device seems brittle in the face of possible future changes.
This choice of `/var/log` rather than `/dev/sda5` is relevant regardless of whether we use `df` or `os.statvfs`.

File: scripts/perfmon [hunk 1]
```diff
@@ -345,11 +345,23 @@ class RRDUpdates:
...
 
 def average(mylist):
     return sum(mylist)/float(len(mylist))
 
+def get_percent_log_fs_usage(ignored):
+    "Get the percent usage of the host filesystem for logs partition. Input list is ignored and should be empty"
+    output = commands.getoutput('df /dev/sda5')
+    if not (re.search(r'No such file or directory', output)):
+        output = ' '.join(output.splitlines()[1:])
+        if (output.split()[5]) == '/var/log':
```
Comment (thomassa) [inline scripts/perfmon]: If this comparison is false, `percentage` will not be set at all, so line 363 will thrown an exception when it attempts `return float(percentage[0:-1])/100.0` with undefined `percentage`.

**Comment by sharady** [comment]
@johnelse, @thomassa  thanks for the comments:)
Using '/dev/sda5' looks brittle i know, But I made it under impression, If there is a case when log partition is no different than main filesystem partition. In that case `df /var/log` will point to the main filesystem space usage.
In the patch, I am cross checking '/dev/sda5' is present and mounted to '/var/log'.
Any suggestions are welcome :)

File: scripts/perfmon (hunk 1 above)
Comment (sharady) [inline scripts/perfmon]: Nice catch will update that, thanks :)

**Comment by thomassa** [comment]
> If there is a case when log partition is no different than main filesystem partition. In that case `df /var/log` will point to the main filesystem space usage.

That is true, but I think it is reasonable to report the percentage in that case anyway.

If I were to be persuaded that such a case should be treated differently, then I would recommend detecting it by comparing the devices on which `/var/log` and `/etc/passwd` are mounted to see if they are the same: there would still be no need for perfmon to know about `/dev/sda5`. (You can check the device where a path is mounted by calling https://docs.python.org/2/library/os.html#os.stat and examining `st_dev` in the result. Alternatively, it is the first item on the second line of the output of `df`.)

**Comment by sharady** [comment]
Using `df` in new function `get_percent_log_fs_usage`, to keep it consistent with existing function `get_percent_fs_usage`.
`os.stat.st_dev` doesn't provide any useful information, relying on `df` to find the path mounted.
Will squash the commits when changes looks good.

**Comment by thomassa** [comment]
When the logs are on the main partition, I suppose it's to avoid duplicate alarms that you're making the function claim the occupy 0% rather than the truth. It's potentially misleading though. How about returning Not-a-Number instead?
i.e. instead of the line
`       percentage = '0%'` it would be
`       return float('NaN')`

File: scripts/perfmon [hunk 2]
```diff
@@ -345,11 +345,25 @@ class RRDUpdates:
...
+    fs_output = commands.getoutput('df /etc/passwd')
+    log_fs_output = commands.getoutput('df /var/log')
+    fs_output = ' '.join(fs_output.splitlines()[1:])
+    log_fs_output = ' '.join(log_fs_output.splitlines()[1:])
+    # Get the percent usage only when there is a separate logs partition
+    if (fs_output.split()[0] != log_fs_output.split()[0]):
+        percentage = log_fs_output.split()[4]
+    else:
+        percentage = '0%'
+    # strip of % character and convert to float
```
Comment (thomassa) [inline scripts/perfmon]: Typo: "of" should be "off".
(Also in get_percent_fs_usage.)

**Comment by thomassa** [comment]
The code looks fine now, so the commits need squashing into one and then it should be good to merge.

**Comment by thomassa** [comment]
In it goes.

**Pull request merged**